vyper/builtin_functions/functions.py

```diff
@@ -2156,8 +2158,69 @@ def build_IR(self, expr, context):
         )
 
 
+class ABIDecode(_SimpleBuiltinFunction):
+    _id = "_abi_decode"
+    _inputs = [("data", BytesArrayPrimitive())]
+
+    def fetch_call_return(self, node):
+        self.infer_arg_types(node)
+
+        # Get parent node for type inference
+        parent_node = node.get_ancestor()
+        if isinstance(parent_node, vy_ast.Assign):
```

Comment:
hm, what if this is an `AnnAssign`? i'd kind of prefer this ad-hoc logic to be moved into the typechecker / annotation phase

Reply:
Right, `AnnAssign` needs to be added as well.

Moving this into typechecker / annotation phase at this point seems awkward. Since annotation is after type-checking, we cannot validate that `_abi_decode` (RHS) against the expected type (LHS). (Currently, in the typechecking phase (see `visit_AnnAssign` of `semantics/validation/local.py`), there is a call to `validate_expected_type` to check if the RHS can be of the LHS type. Since `_abi_decode` is a `Call` node, this calls back to `fetch_call_return` which depends on annotation but annotation runs after type-checking.)

Reply:
the issue is we are recreating logic from the typechecker. and we might do it incorrectly, too. what if `_abi_decode()` occurs inside of a Call node, or some other expression?

if we do move to typechecker/annotation we will definitely need ad-hoc rules to deal with `_abi_decode`, but it will be in one place instead of many. for instance, we can just have the output of `_abi_decode()` always typecheck.

Reply:
Resolved since we are no longer deriving output type from LHS.